components/raftstore-v2/src/operation/command/admin/leader_transfer.rs

```diff
@@ -0,0 +1,419 @@
+    ///     does, it calls raft transfer_leader API to do the remaining work.
+    ///
+    /// Additional steps when there are remaining pessimistic
+    /// locks to propose (detected in function on_transfer_leader_msg).
+    ///    1. Leader firstly proposes pessimistic locks and then proposes a
+    ///       TransferLeader command.
+    ///    2. The follower applies the TransferLeader command and replies an
+    ///       ACK with special context TRANSFER_LEADER_COMMAND_REPLY_CTX.
+    ///
+    /// See also: tikv/rfcs#37.
+    pub fn propose_transfer_leader<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+        ch: CmdResChannel,
+    ) -> bool {
+        ctx.raft_metrics.propose.transfer_leader.inc();
+
+        let transfer_leader = get_transfer_leader_cmd(&req).unwrap();
+        let prs = self.raft_group().raft.prs();
+
+        // Find the target with the largest matched index among the candidate
+        // transferee peers
+        let (_, peers) = transfer_leader
+            .get_peers()
+            .iter()
+            .filter(|peer| peer.id != self.peer().id)
+            .fold((0, vec![]), |(max_matched, mut chosen), p| {
+                if let Some(pr) = prs.get(p.id) {
+                    match pr.matched.cmp(&max_matched) {
+                        Ordering::Greater => (pr.matched, vec![p]),
+                        Ordering::Equal => {
+                            chosen.push(p);
+                            (max_matched, chosen)
+                        }
+                        Ordering::Less => (max_matched, chosen),
+                    }
+                } else {
+                    (max_matched, chosen)
+                }
+            });
+        let peer = match peers.len() {
+            0 => transfer_leader.get_peer(),
+            1 => peers.get(0).unwrap(),
+            _ => peers.choose(&mut rand::thread_rng()).unwrap(),
+        };
+
+        let transferee = if peer.id == self.peer().id {
+            false
+        } else {
+            self.pre_transfer_leader(peer)
+        };
+
+        // transfer leader command doesn't need to replicate log and apply, so we
+        // return immediately. Note that this command may fail, we can view it just as
+        // an advice
+        ch.set_result(make_transfer_leader_response());
+
+        transferee
+    }
+
+    fn pre_transfer_leader(&mut self, peer: &metapb::Peer) -> bool {
+        if self.raft_group().raft.has_pending_conf() {
+            info!(
+                self.logger,
+                "reject transfer leader due to pending conf change";
+                "peer" => ?peer,
+            );
+            return false;
+        }
+
+        // Broadcast heartbeat to make sure followers commit the entries immediately.
+        // It's only necessary to ping the target peer, but ping all for simplicity.
+        self.raft_group_mut().ping();
+
+        let mut msg = eraftpb::Message::new();
```

Comment:
Here, the index is missing compared to v1. I think it is worth a TODO comment because the index is important for reducing jitter.

Reply:
Yes, that's the warm up optimization which should be implemented later. I will add todo.